agent/xds/listeners.go

```diff
@@ -270,6 +354,46 @@ func (s *Server) makePublicListener(cfgSnap *proxycfg.ConfigSnapshot, token stri
 	return l, err
 }
 
+func (s *Server) makeExposedCheckListener(cfgSnap *proxycfg.ConfigSnapshot, cluster, path, protocol string, port int) (proto.Message, error) {
+	cfg, err := ParseProxyConfig(cfgSnap.Proxy.Config)
+	if err != nil {
+		// Don't hard fail on a config typo, just warn. The parse func returns
+		// default config if there is an error so it's safe to continue.
+		s.Logger.Printf("[WARN] envoy: failed to parse Connect.Proxy.Config: %v", err)
+	}
+
+	// No user config, use default listener
+	addr := cfgSnap.Address
+
+	// Override with bind address if one is set, otherwise default to 0.0.0.0
+	if cfg.BindAddress != "" {
+		addr = cfg.BindAddress
+	} else if addr == "" {
+		addr = "0.0.0.0"
+	}
+
+	// Strip any special characters from path
+	r := regexp.MustCompile(`[^a-zA-Z0-9]+`)
+	strippedPath := r.ReplaceAllString(path, "")
+	listenerName := fmt.Sprintf("exposed_path_%s", strippedPath)
```

Comment:
🤔 I might have had the wrong mental model but this explains my comment above about grouping paths by port. Having separate listener for each path though _requires_ a service to listen on multiple ports if they want to expose multiple checks. I think people will quickly find that painful.

Was this a simplification for now? Do you intend to allow a single listener/port in future that can expose multiple paths in the app? That seems desirable as many apps will have at minimum a `/healthz` and a `metrics` endpoint. Exposing each as a separate port is OK but a bit wasteful/unintuitive.

Reply:
I'm not sure that I follow here. Having a separate listener for each path doesn't require the service to listen on multiple ports, since listeners for different ports can point to the same downstream cluster.

Reply:
No but it requires Envoy to listen on multiple ports - each listener must have a unique bind address:port pair. So for example if my service has `/healthz` and `/statsz` endpoints to expose I'd end up with two separate listeners on different ports for them instead of one that can access either.

I don't feel super strongly as it works this way and is maybe not worth the complication to combine them but it was surprising to me that we wouldn't at least try to reduce to just one extra port to open up on firewalls etc per proxy for this rather than one per endpoint exposed.

Reply:
Ah, I see. I can make an issue to later combine ports when possible. We would be able to create listeners for each protocol/TLS context combination. 

Another concern I had about that is that I'm not sure if it's a weird experience to have to assign unique ports on a per-protocol/CACert basis. But we can add that to the NodeService validation to make that easier to set up correctly.

Reply:
Oh I just remembered another concern about that. We don't know whether Consul's checks are HTTP2 or not (and Envoy currently can't figure that out on its own). We would need to expand the check definition to include that distinction. Or we could consolidate ports by protocol when exposing paths, but not when exposing checks.

Reply:
Deleted the last comment, quickly realized it was a non-issue.